### ISSUE TITLE
Model#save() is not returning plugged in Promise

### ISSUE THREAD

**Issue by AndrewBarba** [Automattic/mongoose#5030]
**What is the current behavior?**
When calling `doc.save()` I am getting back a promise that was not created using the constructor I specified.

**What is the expected behavior?**
When calling `doc.save()` I should have a Promise of the same type that I specified in `mongoose.Promise = ...`.

**Please mention your node.js, mongoose and MongoDB version.**
Node 6.10.0, Mongoose 4.8.5 (Upgrading from 4.6.x)

**Comment by johnny88** [comment]
@AndrewBarba  So the problem is that you are not calling mongoose.Promise = whatever in the right place. It's kind of hard to figure out the right place since the docs don't really specify. Is there anyway I could look at the code?

**Comment by AndrewBarba** [comment]
@johnny88 following up from other ticket, where is correct place to set the Promise library? We have quite a large code base, would be difficult to share. But to oversimplify this, we have a root `server.js` file, and that file simply requires all the Models. First line in `server.js` I have:

```
const Promise = require('bluebird')
const mongoose = require('mongoose');
mongoose.Promise = Promise;
```

Can you shed any light on why setting at that early would make a difference? Should it be set later in the process?

**Comment by johnny88** [comment]
@AndrewBarba Hey, you pretty much have to call the promise every where you import mongoose, not just the one place. I presume that the server.js only connects to the mongo db through mongoose and does not actually define the schema. I would through the mongoose Promise into your schema and see if that works, it worked for me.

Unfortunately since I am not a mongoose developer only a user who has experienced the same issue I couldn't tell you why this is. Hope this helps!

**Comment by varunjayaraman** [comment]
I reset the mongoose Promise constructor in every project that I use and I never get the deprecation warning. @AndrewBarba the reason you have to set it as early as possible is that `require()` in node is a singleton, so by setting the promise constructor early on in the process, it will mutate `mongoose.Promise` and use the promise you've set everywhere else in the project.

**Comment by AndrewBarba** [comment]
@varunjayaraman Thats my exact issue though, I am setting as early as possible, thats what I expected, and it is still not working... Somewhere it is getting reset back

**Comment by johnny88** [comment]
@AndrewBarba  @varunjayaraman I figured it was getting reset on every import. I logged out the files and it seem when I import/required a second time it would reset it back. That's why I went into each file and set the mongoose.Promise again. It fixed my problem at the expense of a little duplication. Something I was hoping would disappear when they implemented native promises in mongoose. :)

I don't think setting it again would cause any problems, but it would be a lot nicer and cleaner to be able to just set it once.

**Comment by AndrewBarba** [comment]
To shed a little more light, I just added a file to the root of my project at `local-mongoose.js`, it looks like this:

```
const mongoose = require('mongoose');

mongoose.Promise = require('bluebird');

module.exports = mongoose;
```

I replaced every instance of `require('mongoose')` in our code base with `require('local-mongoose')` and now everything works as expected. So something funky going on here. Didn't have to do this in any previous version of Mongoose. @johnny88 doing it this way gets rid of the duplication, but not ideal...

**Comment by varunjayaraman** [comment]
@AndrewBarba yeah that's one way to do it. I admit it's a bit wonky to get around this bug because completely removing mpromise is a major version breaking change, but at the same time we need some way to enable users to use native Promises and to share that config through the mongoose singleton.

**Comment by AndrewBarba** [comment]
@varunjayaraman @vkarpov15 can we tag this as a confirmed bug? It's clear that resetting `mongoose.Promise = ...` after every require fixes the issue, and that something internally is resetting this. We should only have to call that once, as early in the app lifecycle as possible. Would be great to track this and look at getting it fixed. I think it's a different issue than just removing mpromise. If we set the Promise ourself I would expect it to not be reset later on

**Comment by vkarpov15** [comment]
@AndrewBarba it's not a confirmed bug because I don't have any way to reproduce this. You should only set `mongoose.Promise` once, if that doesn't work it's a bug. Please provide repro instructions, OP is a classic example of a terrible bug report.

**Comment by VitoVan** [comment]
@vkarpov15 
Here's a way to reproduce this:
https://github.com/thebergamo/k7-mongoose/issues/6#issuecomment-282193140

If it is not clear, I can provide another test case.

**Comment by AndrewBarba** [comment]
Hahaha terrible report. Have you read the other reports in your repo? I filled out YOUR template. If it's terrible, change the template.

**Comment by varunjayaraman** [comment]
@AndrewBarba in all fairness, you didn't answer half the questions from the template, and the questions you omitted are the most important, i.e. how to reproduce. Don't criticize the template if you are barely going to even use it

**Comment by AndrewBarba** [comment]
In all fairness, there is real discussion in this thread. Code samples, solutions, plenty of detail. To call the issue "terrible" is a great way to discourage anyone else from reporting things in the first place. If you absolutely will not move forward until the issue needs a clear project reproducing it then change the tag from needs clarification to something more useful to me. With a 7+ day response time it's hard for me to know what you need next.

**Comment by AndrewBarba** [comment]
And while we're at it, have a look at your template. There's 4 questions. I answered 3. Last I checked that's more than half, but I'll forgive your hyperbole. I can't always give exact steps to reproduce until we get some discussion going. The template isn't even close to standard, look at Realm Fastlane etc. to see a real template. Not to mention 99% of the issues in this repo don't even answer a single question in the template and remove it entirely. You guys have multiple issues that mention the deprecation warning and people experiencing similar things to myself. It's all the same issue. Something is resetting the Promise library. I said this. I said change the name to help you guys organize better. I didn't ask for an immediate fix I just asked it be classified as a bug since we have so much detail explaining what's going on. Needs clarification means this is going to get lost in the weeds, I want to avoid that.

**Comment by varunjayaraman** [comment]
Alright @AndrewBarba so it's difficult to tag it as a confirmed-bug if there's no clearcut way to reproduce it. I've created a sample repo here: https://github.com/varunjayaraman/mongoose-promise-bug thought I thought people might be able to experiment with to try and reproduce the bug. The repo is pretty barebones and, at least for me, does **not** produce the warning.

Hopefully this will make it easier to debug since it seems most people's apps are a bit too large to try and reproduce. 

@VitoVan the `save()` test actually works for me in the repo you provided. I commented out the `defaultPromise` setting and it fails, which it actually **should** do in the example you provided. Here's why: You've already called `mongoose.model()` before you set the `defaultPromise` because you required in the `CandyModel` at the top of the file, and that file calls `mongoose.model()`. So the promise constructor has _already_ been set for instances of the `CandyModel`. You **must** set `mongoose.Promise` before you call `mongoose.model()`.

**Comment by varunjayaraman** [comment]
Also tracking the following issues from here:

https://github.com/Automattic/mongoose/issues/5020
https://github.com/Automattic/mongoose/issues/4951
https://github.com/Automattic/mongoose/issues/4943

**Comment by VitoVan** [comment]
@varunjayaraman 

As you say: 

> So the promise constructor has already been set for instances of the CandyModel. You must set mongoose.Promise before you call mongoose.model().

Then, this is **NOT A BUG**, it's **SUPPOSED SO**. 

#### For anybody who still consider this as a bug, here's a more clear explanation about WHY:

Promise constructor is set **ONCE YOU CALL** `mongoose.model()`

#### And here's the ugly solution:

https://github.com/Automattic/mongoose/issues/5030#issuecomment-283395174

> added a file to the root of my project called local-mongoose.js

and in there, you set `mongoose.Promise`, and use `require('local-mongoose')` instead of `require('mongoose')`.

#### And here's the future:

https://github.com/Automattic/mongoose/issues/5030#issuecomment-283395586

> ......completely removing mpromise is a major version breaking change......


Well, this is it.

**Comment by varunjayaraman** [comment]
I labeled it a `bug?` because so many people have reported errors about it. I'm often suspicious that the problem stems from people importing mongoose after any `mongoose.model()` calls but without full access to their repository, it's not easy to debug those problems.

I'm personally not able to reproduce it in any projects where mongoose **is** imported properly (i.e. before any and all attempts to require in files that use mongoose).

If anyone is able to reproduce the issue with the repo i posted above, I'd love to take a look and see if it's actually a problem with mongoose or a flaw in the way mongoose was imported

**Comment by thebergamo** [comment]
@varunjayaraman check this: https://github.com/thebergamo/k7-mongoose/tree/feat-default-promise

In this branch of my repo, I'm trying to configure the correct promise in the constructor of the class. When I try to use the `findOne()`, that's ok. But trying the `save()` the returned promise yet the `mpromise` instance.

the tests about the feature: https://github.com/thebergamo/k7-mongoose/blob/feat-default-promise/test/index.js#L135-L203

**Comment by vkarpov15** [comment]
@thebergamo thanks for the repro. It's an issue where, for `save()`, mongoose will currently use the promise constructor that it was configured to use at the time the model was created. It looks like your models were already created before the test set `mongoose.Promise`. In general, I'd be careful with your overall paradigm of interacting with mongoose: `mongoose.model()` attaches models to the mongoose singleton, so having an object that you create with `new` that wraps a singleton sounds dangerous.

Either way, your tests pass with the above fix, the fix will be in 4.9.1, ETA Sunday March 19

**Comment by thebergamo** [comment]
Yeah, its the way I find to use the same singleton provided by my module. But I trying to find a new way to do this. In the future probably will change. But I export an static method instead of a instance method.

Will update the mongoose on my module to support this feature. Thanks very much @vkarpov15
